File: atcoder/abc/abc037/d.rb
# frozen_string_literal: true

$h, $w = gets.split.map(&:to_i)
$a = $h.times.map { gets.split.map(&:to_i) }.flatten
MOD = 1_000_000_007

$dp = Array.new($h * $w,-1)

def rec(y, x)
  return $dp[y * $w + x] if $dp[y * $w + x] != -1

  ret = 1
  ayx = $a[y * $w + x]
  if x - 1 >= 0 && $a[y * $w + x - 1] < ayx
    ret += rec(y, x - 1)
    ret %= MOD if ret >= MOD
  end
  if x + 1 < $w && $a[y * $w + x + 1] < ayx
    ret += rec(y, x + 1)
    ret %= MOD if ret >= MOD
  end
  if y - 1 >= 0 && $a[(y - 1) * $w + x] < ayx
    ret += rec(y - 1, x)
    ret %= MOD if ret >= MOD
  end
  if y + 1 < $h && $a[(y + 1) * $w + x] < ayx
    ret += rec(y + 1, x)
    ret %= MOD if ret >= MOD
  end

  $dp[y * $w + x] = ret
end

ans = 0

$h.times do |i|
  $w.times do |j|
    ans += rec(i, j)
    ans %= MOD if ans >= MOD
  end
end
p ans

File: atcoder/abc/abc037/d.cpp
#include <bits/stdc++.h>

#define rep(i, n) for (int i = 0; i < (int)(n); i++)
#define all(x) (x).begin(), (x).end()

using namespace std;
using ll = int64_t;

constexpr int INF = (1 << 30);
constexpr ll INFL = (1LL << 62);
constexpr ll MOD = 1000000007;
int dy[] = {1, 0, -1, 0};
int dx[] = {0, 1, 0, -1};

int a[1010][1010];
ll dp[1010][1010];
int h, w;
ll rec(int y, int x) {
    if (dp[y][x] != -1) return dp[y][x];
    ll ret = 1;
    rep(i, 4) {
        int ny = y + dy[i], nx = x + dx[i];
        if (!(0 <= ny && ny < h && 0 <= nx && nx < w)) continue;
        if (a[y][x] < a[ny][nx]) {
            ret += rec(ny, nx);
            ret %= MOD;
        }
    }

    return dp[y][x] = ret;
}

int main() {
    cout << std::fixed << std::setprecision(10);
    cin >> h >> w;
    rep(i, h) rep(j, w) scanf("%d", &a[i][j]);
    memset(dp, -1, sizeof(dp));
    ll ans = 0;
    rep(i, h) rep(j, w)(ans += rec(i, j)) %= MOD;
    cout << ans << endl;
    return 0;
}
